compiler/rustc_mir_dataflow/src/value_analysis.rs

```diff
@@ -0,0 +1,677 @@
+        _location: Location,
+    ) {
+        self.0.handle_statement(statement, state);
+    }
+
+    fn apply_terminator_effect(
+        &self,
+        state: &mut Self::Domain,
+        terminator: &Terminator<'tcx>,
+        _location: Location,
+    ) {
+        self.0.handle_terminator(terminator, state);
+    }
+
+    fn apply_call_return_effect(
+        &self,
+        state: &mut Self::Domain,
+        _block: BasicBlock,
+        return_places: crate::CallReturnPlaces<'_, 'tcx>,
+    ) {
+        self.0.handle_call_return(return_places, state)
+    }
+
+    fn apply_switch_int_edge_effects(
+        &self,
+        _block: BasicBlock,
+        discr: &Operand<'tcx>,
+        apply_edge_effects: &mut impl SwitchIntEdgeEffects<Self::Domain>,
+    ) {
+        self.0.handle_switch_int(discr, apply_edge_effects)
+    }
+}
+
+rustc_index::newtype_index!(
+    pub struct PlaceIndex {}
+);
+
+rustc_index::newtype_index!(
+    struct ValueIndex {}
+);
+
+#[derive(PartialEq, Eq, Clone, Debug)]
+pub struct State<V>(IndexVec<ValueIndex, V>);
+
+impl<V: Clone + HasTop> State<V> {
+    pub fn flood_all(&mut self, value: V) {
+        self.0.raw.fill(value);
+    }
+
+    pub fn flood(&mut self, place: PlaceRef<'_>, map: &Map, value: V) {
+        if let Some(root) = map.find(place) {
+            self.flood_idx(root, map, value);
+        }
+    }
+
+    pub fn flood_idx(&mut self, place: PlaceIndex, map: &Map, value: V) {
+        map.preorder_invoke(place, &mut |place| {
+            if let Some(vi) = map.places[place].value_index {
+                self.0[vi] = value.clone();
+            }
+        });
+    }
+
+    pub fn assign_place_idx(&mut self, target: PlaceIndex, source: PlaceIndex, map: &Map) {
+        if let Some(target_value) = map.places[target].value_index {
+            if let Some(source_value) = map.places[source].value_index {
+                self.0[target_value] = self.0[source_value].clone();
+            } else {
+                self.0[target_value] = V::top();
+            }
+        }
+        for target_child in map.children(target) {
+            // Try to find corresponding child in source.
+            let projection = map.places[target_child].proj_elem.unwrap();
+            if let Some(source_child) = map.projections.get(&(source, projection)) {
+                self.assign_place_idx(target_child, *source_child, map);
+            } else {
+                self.flood_idx(target_child, map, V::top());
+            }
+        }
+    }
+
+    pub fn assign(&mut self, target: PlaceRef<'_>, result: ValueOrPlace<V>, map: &Map) {
+        if let Some(target) = map.find(target) {
+            self.assign_idx(target, result, map);
+        }
+    }
+
+    pub fn assign_idx(&mut self, target: PlaceIndex, result: ValueOrPlace<V>, map: &Map) {
+        match result {
+            ValueOrPlace::Value(value) => {
+                // FIXME: What if not all tracked projections are overwritten? Can this happen?
```

Comment:
At the very least in case of deref projections, absolutely. The code below unfortunately has a bunch of extra `&mut *` things, but I believe it would miscompile if those were gone.

```rust
#[inline(never)]
fn foo() -> usize {
    static mut P: usize = 5;

    let mut x = 0;
    let mut r = &mut x;
    *r = 0;
    r = unsafe { &mut P };
    *r
}

fn main() {
    dbg!(foo());
}
```

Reply:
I think you are right, the target place must be flooded. However, I want to clarify that this example would not break as-is, because the problematic assignment `r = unsafe { &mut P };` would yield `ValueOrPlace::Unknown`, which would flood `r` and therefore `*r` anyway. However, a different analysis would perhaps track values for the references themselves, which would require the invalidation also for `ValueOrPlace::Value`.

Reply:
> because the problematic assignment `r = unsafe { &mut P };` would yield `ValueOrPlace::Unknown`

Why would it yield `Unknown`? `unsafe { &mut P }` is a constant

Reply:
Oh, now I see what you mean. You are absolutely right. Modifying your example a little bit yields the following program, which breaks the analysis completely.
```rust
fn main() {
    static P: usize = 5;
    let x = 0;
    let mut r = &x;
    r = &P;
    let y = *r;
}
```
BEFORE and AFTER adding additional flooding:

<img src="https://user-images.githubusercontent.com/30163784/187693036-d3e92e60-46ee-4b22-b24b-5c3f90ae2bc6.png" height="600px"/><img src="https://user-images.githubusercontent.com/30163784/187693404-cbe36e75-8e70-42ec-b764-62e9cfaee7d3.png" height="600px"/>

So this will be fixed with the new commit. Thank you so much!